aws/resource_aws_ec2_traffic_mirror_filter_rule.go

```diff
@@ -0,0 +1,370 @@
+	}
+
+	d.SetId(*out.TrafficMirrorFilterRule.TrafficMirrorFilterRuleId)
+	return resourceAwsEc2TrafficMirrorFilterRuleRead(d, meta)
+}
+
+func resourceAwsEc2TrafficMirrorFilterRuleRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).ec2conn
+	ruleId := d.Id()
+
+	var rule *ec2.TrafficMirrorFilterRule
+	filterId, filterIdSet := d.GetOk("traffic_mirror_filter_id")
+	input := &ec2.DescribeTrafficMirrorFiltersInput{}
+	if filterIdSet {
+		input.TrafficMirrorFilterIds = []*string{aws.String(filterId.(string))}
+		awsMutexKV.Lock(filterId.(string))
+		defer awsMutexKV.Unlock(filterId.(string))
+	} else {
+		awsMutexKV.Lock("DescribeTrafficMirrorFilters")
+		defer awsMutexKV.Unlock("DescribeTrafficMirrorFilters")
+	}
+
+	found := false
+	for !found {
+		out, err := conn.DescribeTrafficMirrorFilters(input)
+		if err != nil {
+			return fmt.Errorf("Error listing traffic mirror filters in the account")
+		}
+
+		if 0 == len(out.TrafficMirrorFilters) {
+			return fmt.Errorf("No traffir mirror filters found")
+		}
+
+		_, rule, err = findRuleInFilters(ruleId, out.TrafficMirrorFilters)
+		if nil == err {
+			found = true
+			break
+		} else {
+			if out.NextToken == nil {
+				break
+			} else {
+				input.NextToken = out.NextToken
+			}
+		}
+	}
+
+	if !found {
+		d.SetId("")
+		return fmt.Errorf("Rule %s not found", ruleId)
+	}
+
+	return populateTrafficMirrorFilterRuleResource(d, rule)
+}
+
+func findRuleInFilters(ruleId string, filters []*ec2.TrafficMirrorFilter) (filter *ec2.TrafficMirrorFilter, rule *ec2.TrafficMirrorFilterRule, err error) {
+	log.Printf("[DEBUG] searching %s in %d filters", ruleId, len(filters))
+	err = nil
+	found := false
+	for _, v := range filters {
+		log.Printf("[DEBUG]: searching filter %s, ingress rule count = %d, egress rule count = %d", *v.TrafficMirrorFilterId, len(v.IngressFilterRules), len(v.EgressFilterRules))
+		for _, r := range v.IngressFilterRules {
+			if *r.TrafficMirrorFilterRuleId == ruleId {
+				rule = r
+				filter = v
+				found = true
+				break
+			}
+		}
+		for _, r := range v.EgressFilterRules {
+			if *r.TrafficMirrorFilterRuleId == ruleId {
+				rule = r
+				filter = v
+				found = true
+				break
+			}
+		}
+		if found {
+			break
+		}
+	}
+
+	if !found {
+		err = fmt.Errorf("Rule %s not found", ruleId)
+	}
+
+	log.Printf("[DEBUG]: Found %s in %s %s", ruleId, *filter.TrafficMirrorFilterId, *rule.TrafficDirection)
+	return filter, rule, err
+}
+
+func resourceAwsEc2TrafficMirrorFilterRuleUpdate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).ec2conn
+	ruleId := d.Id()
+
+	input := &ec2.ModifyTrafficMirrorFilterRuleInput{
+		TrafficMirrorFilterRuleId: &ruleId,
+	}
+
+	var removeFields []*string
+	if d.HasChange("protocol") {
+		_, n := d.GetChange("protocol")
+		if n == "0" {
+			removeFields = append(removeFields, aws.String("protocol"))
+		} else {
+			input.Protocol = aws.Int64(int64(d.Get("protocol").(int)))
+		}
```

Comment:
Only using the second return value of `d.GetChange()` is equivalent to `d.Get()`, so would recommend the following (along with AWS Go SDK constants) here and below in the rest of the Update functions:

```go
		v := d.Get("protocol").(int)
		if v == 0 {
			removeFields = append(removeFields, aws.String(ec2.TrafficMirrorFilterRuleFieldProtocol))
		} else {
			input.Protocol = aws.Int64(int64(v))
		}
```

See also: https://github.com/bflad/tfproviderlint/issues/44